arch/xtensa/src/esp32/esp32_psram.c

```diff
@@ -0,0 +1,1693 @@
+#define PSRAM_RESET                0x99
+#define PSRAM_SET_BURST_LEN        0xc0
+#define PSRAM_DEVICE_ID            0x9f
+
+typedef enum
+{
+  PSRAM_CLK_MODE_NORM = 0,    /* Normal SPI mode */
+  PSRAM_CLK_MODE_DCLK = 1,    /* 2 clock cycles after CS is high level */
+} psram_clk_mode_t;
+
+#define PSRAM_ID_KGD_M             0xff
+#define PSRAM_ID_KGD_S             8
+#define PSRAM_ID_KGD               0x5d
+#define PSRAM_ID_EID_M             0xff
+#define PSRAM_ID_EID_S             16
+
+/* Use the [7:5](bit7~bit5) of EID to distinguish the psram size:
+ *
+ *   BIT7  |  BIT6  |  BIT5  |  SIZE(MBIT)
+ *   -------------------------------------
+ *    0    |   0    |   0    |     16
+ *    0    |   0    |   1    |     32
+ *    0    |   1    |   0    |     64
+ */
+
+#define PSRAM_EID_SIZE_M           0x07
+#define PSRAM_EID_SIZE_S           5
+
+typedef enum
+{
+  PSRAM_EID_SIZE_16MBITS = 0,
+  PSRAM_EID_SIZE_32MBITS = 1,
+  PSRAM_EID_SIZE_64MBITS = 2,
+} psram_eid_size_t;
+
+#define PSRAM_KGD(id)         (((id) >> PSRAM_ID_KGD_S) & PSRAM_ID_KGD_M)
+#define PSRAM_EID(id)         (((id) >> PSRAM_ID_EID_S) & PSRAM_ID_EID_M)
+#define PSRAM_SIZE_ID(id)     ((PSRAM_EID(id) >> PSRAM_EID_SIZE_S) & PSRAM_EID_SIZE_M)
+#define PSRAM_IS_VALID(id)    (PSRAM_KGD(id) == PSRAM_ID_KGD)
+
+/* For the old version 32Mbit psram, using the special driver */
+
+#define PSRAM_IS_32MBIT_VER0(id)  (PSRAM_EID(id) == 0x20)
+#define PSRAM_IS_64MBIT_TRIAL(id) (PSRAM_EID(id) == 0x26)
+
+/* IO-pins for PSRAM.
+ * WARNING: PSRAM shares all but the CS and CLK pins with the flash, so these
+ * defines hardcode the flash pins as well, making this code incompatible
+ * with either a setup that has the flash on non-standard pins or ESP32s with
+ * built-in flash.
+ */
+
+#define PSRAM_SPIQ_SD0_IO          7
+#define PSRAM_SPID_SD1_IO          8
+#define PSRAM_SPIWP_SD3_IO         10
+#define PSRAM_SPIHD_SD2_IO         9
+
+#define FLASH_HSPI_CLK_IO          14
+#define FLASH_HSPI_CS_IO           15
+#define PSRAM_HSPI_SPIQ_SD0_IO     12
+#define PSRAM_HSPI_SPID_SD1_IO     13
+#define PSRAM_HSPI_SPIWP_SD3_IO    2
+#define PSRAM_HSPI_SPIHD_SD2_IO    4
+
+/* PSRAM clock and cs IO should be configured based on hardware design.
+ * For ESP32-WROVER or ESP32-WROVER-B module, the clock IO is IO17, the CS IO
+ * is IO16, they are the default value for these two configs.
+ */
+
+#define D0WD_PSRAM_CLK_IO          CONFIG_D0WD_PSRAM_CLK_IO  /* Default is 17 */
+#define D0WD_PSRAM_CS_IO           CONFIG_D0WD_PSRAM_CS_IO   /* Default is 16 */
+
+#define D2WD_PSRAM_CLK_IO          CONFIG_D2WD_PSRAM_CLK_IO  /* Default is 9 */
+#define D2WD_PSRAM_CS_IO           CONFIG_D2WD_PSRAM_CS_IO   /* Default is 10 */
+
+/* For ESP32-PICO chip, the psram share clock with flash. The flash clock
+ * pin is fixed, which is IO6.
+ */
+
+#define PICO_PSRAM_CLK_IO          6
+#define PICO_PSRAM_CS_IO           CONFIG_PICO_PSRAM_CS_IO   /* Default is 10 */
+
+typedef struct
+{
```

Comment:
This is an example of the sections I was referring to.  Structures and enumerations should be in their own Private Types section.